### PR TITLE
#12388 Remove TestCase.defer{SetUp, TestMethod, TearDown, RunCleanups}

### LINKED ISSUE #12388: Remove TestCase.deferSetUp, TestCase.deferTestMethod, TestCase.deferTearDown and TestCase.deferRunCleanups
This issue has been discovered in https://github.com/twisted/twisted/pull/12375.

Currently `TestCase` has the following functions that are presumably public due to their naming: `deferSetUp`, `deferTestMethod`, `deferRunCleanups` and `deferTearDown`. This raises a few problems. I suggest removing the functions without deprecation period due to the following reasons:

 - these functions expose so much of the internal structure of TestCase that it makes the code hard to change.
 - there is no case where the functions would be useful as an extension point. The only use of the functions that I found using Github code search is my own code in the Buildbot project which overrides `deferRunCleanups` and hooks even deeper into TestCase internals.
 - the API is not "public ready" (i.e. the `ignored` parameter) - it seems as if they were some kind of internal thing that simply didn't get the `_` prefix.
 - deprecation would be harder to implement than usual, due to the functions being supposed to be overridden, not just called.
 - this is API for tests, so it is not likely to break functionality that is used in production. Only development workflows may be affected where the fix is to update the tests.

Given the above - small risk of breaking valid code for someone and relatively high costs for going through the deprecation dance I would suggest to have an exception for this change and allow to remove the functions without deprecation period.

### PR DESCRIPTION
**Pull request by p12tic**
Fixes https://github.com/twisted/twisted/issues/12388.

### CODE REVIEW

**Comment by p12tic** [comment]
please review

**Comment by p12tic** [comment]
I've sent an email to the twisted mailing list at twisted@python.org, but it seems that my mail didn't get through yet.

**Comment by graingert** [comment]
> I've sent an email to the twisted mailing list at [twisted@python.org](mailto:twisted@python.org), but it seems that my mail didn't get through yet.

You could try using the web interface: https://mail.python.org/archives/list/twisted@python.org/message/new

**Comment by p12tic** [comment]
I think my mails could be in moderator queue. Hopefully the moderator rejects all of them except the most recent one.

**Comment by p12tic** [comment]
Given that https://github.com/twisted/twisted/pull/12375 has been closed, the primary benefit of this PR has vanished. So if my emails don't get through to the mailing list it makes sense to close this PR as well.

**Comment by glyph** [comment]
> I think my mails could be in moderator queue. Hopefully the moderator rejects all of them except the most recent one.

Yep. We have a very aggressive moderator queue where each new member must be manually approved. A *lot* of spammers target this list, and this has been a very effective strategy; they post once, it gets stuck, and they disappear.

Your message has been approved and your email has been set to default moderation, so you should not need to wait in the future.

File: src/twisted/newsfragments/12388.removal [hunk 1]
```diff
@@ -0,0 +1 @@
+Remove ``TestCase.deferSetUp``, ``TestCase.deferTestMethod``, ``TestCase.deferTearDown`` and ``TestCase.deferRunCleanups``.
```
Comment (adiroiban) [inline src/twisted/newsfragments/12388.removal]: Tring to get this to our standard release notes format with the canonical name of the targeted API 

Also, add a note about why we did this .

```suggestion
twisted.trial.unittest.TestCase.deferSetUp, twisted.trial.unittest.TestCase.deferTestMethod, twisted.trial.unittest.TestCase.deferTearDown, and twisted.trial.unittest.TestCase.deferRunCleanups were removed and converted to private methods. These functions expose so much of the internal structure of TestCase that it makes the code hard to change.
```

**Review by adiroiban — APPROVED** [review]
Thanks. All good.

**Comment by p12tic** [comment]
@adiroiban Applied your change, thanks.

**Comment by adiroiban** [comment]
This looks good. Thanks again for the cleanup.

We will wait one week to allow for anyone to be able to test the code with this branch.

If there are no reports or no negative reports, we can merge it.

**Comment by p12tic** [comment]
@adiroiban I think this PR can be merged now.

**Comment by adiroiban** [comment]
Thanks for the follow up and thanks for the PR

I have enabled auto merge.
I hope it will land soon

**Pull request merged**